clientconn.go

```diff
@@ -1057,6 +1055,16 @@ func (ac *addrConn) createTransport(backoffNum int, addr resolver.Address, copts
 
 	onClose := func() {
 		close(onCloseCalled)
+
+		ac.mu.Lock()
+		if ac.state == connectivity.Shutdown {
```

Comment:
As discussed, I believe this should only be done in the allowedToReset case below.  Otherwise the state will go:

Connecting -> Transient Failure -> Connecting -> Transient Failure -> ...

for any addrConn with multiple addresses as each dial attempt within the list fails.  We should stay in Connecting until we get to the very end of the list or succeeds.

I think we also should only be doing it if we received the preface - unless we are currently trying the last address in the list.

Reply:
Hey! I just got an update out that I think mostly resolves this issue, but I need to take another pass over it to double check. In the meantime, let me know what you think.